### ISSUE TITLE
retry TX on a deadlock

### ISSUE THREAD

**Issue by ffffwh** [actiontech/dtle#1050]
1992dd6c

**Comment by ffffwh** [comment]
https://dev.mysql.com/doc/relnotes/mysql/5.7/en/news-5-7-26.html

> InnoDB: Two sessions concurrently executing an INSERT ... ON DUPLICATE KEY UPDATE operation generated a deadlock. During partial rollback of a tuple, another session could update it. The fix for this bug reverts fixes for Bug #11758237, Bug #17604730, and Bug #20040791. (Bug #25966845)

**Comment by ffffwh** [comment]
```sql
create schema if not exists deadlock2;
create table deadlock2.t (
  id int primary key auto_increment,
  val1 int,
  val2 int,
  unique key `uk_val` (val1, val2)
);
-- 注: unique key含null值时, 允许重复.

-- session 1.1
begin;
replace into deadlock2.t values (0, NULL, 1);

-- session 2
begin;
replace into deadlock2.t values (0, NULL, 1);

-- session 1.2
replace into deadlock2.t values (0, NULL, 1);
-- 此时session 2发生deadlock
```
